### ISSUE TITLE
Self-referencing KeyPath falsely triggers self_in_property_initialization

### ISSUE THREAD

**Issue by Deanosaurus** [realm/SwiftLint#5835]
### New Issue Checklist

- [X] I've Updated SwiftLint to the latest version.
- [X] I've searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues).

### Bug Description

When using a key path within property initialization (either as a parameter in a function/initializer for another type or as the property itself) I believe SwiftLint is falsely identifying the key path as referencing a function on `NSObject`.

```swift
final class NotActuallyReferencingSelf {

    // This doesn't actually reference `self` but it shows an error "`self` refers to the unapplied `NSObject.self()`"
    let keyPath: Any = \String.self

    // This does *not* show an error.
    let someType: Any = String.self
}
```

The error mentions `NSObject`, but `String` is a `struct` and you can verify this with any type `\Foo.self`.

### Command/Output

```bash
$ swiftlint --strict
```

> /Path/to/the/file/NotActuallyReferencingSelf.swift:10:5: error: Self in Property Initialization Violation: `self` refers to the unapplied `NSObject.self()` method, which is likely not expected; make the variable `lazy` to be able to refer to the current instance or use `ClassName.self` (self_in_property_initialization)

In reference to this case I believe we are using "`ClassName.self`".

### Environment

* SwiftLint version: 0.57.0
* Xcode version: 16.0.0
* Installation method used: Homebrew
* Configuration file: default

Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? If so, paste their
relative paths and respective contents.

- No nested config
